tracer/src/Datadog.Trace/Iast/DefaultTaintedMap.cs

```diff
@@ -0,0 +1,244 @@
+
+    /// <summary>
+    /// Returns the ITaintedObject for the given input object.
+    /// </summary>
+    /// <param name="objectToFind">The object that should be found in the map</param>
+    /// <returns>The retrieved tainted object or null</returns>
+    public ITaintedObject? Get(object objectToFind)
+    {
+        if (objectToFind is null)
+        {
+            return null;
+        }
+
+        _map.TryGetValue(IndexObject(objectToFind), out var entry);
+
+        while (entry != null)
+        {
+            if (objectToFind == entry.Value)
+            {
+                return entry;
+            }
+
+            entry = entry.Next;
+        }
+
+        return null;
+    }
+
+    /// <summary>
+    /// Put a new TaintedObject in the dictionary.
+    /// </summary>
+    /// <param name="entry">Tainted object</param>
+    public void Put(ITaintedObject entry)
+    {
+        if (entry is null || (entry.Value as string == string.Empty))
+        {
+            return;
+        }
+
+        var index = Index(entry.PositiveHashCode);
+
+        if (!IsFlat)
+        {
+            // By default, add the new entry to the head of the chain.
+            // We do not control duplicate entries.
+            _map.TryGetValue(index, out var existingValue);
+            entry.Next = existingValue;
+
+            // We only count the entries if we are not in flat mode
+            Interlocked.Increment(ref _entriesCount);
+        }
+
+        // If we flipped to flat mode:
+        // - Always override elements ignoring chaining.
+        // - Stop updating the estimated size.
+
+        _map[index] = entry;
+
+        if ((entry.PositiveHashCode & PurgeMask) == 0)
+        {
+            Purge();
+        }
+    }
+
+    /// <summary>
+    /// Purge entries that have been garbage collected. Only one concurrent call to this method is
+    /// allowed, further concurrent calls will be ignored.
+    /// </summary>
+    internal void Purge()
+    {
+        // Ensure we enter only once concurrently.
+        lock (_purgingLock)
+        {
+            if (_isPurging)
+            {
+                return;
+            }
+
+            _isPurging = true;
+        }
+
+        try
+        {
+            // Remove GC'd entries.
+            var removedCount = RemoveDeadKeys();
+
+            if (!IsFlat)
+            {
+                // We only count the entries if we are not in flat mode
+                if (Interlocked.Add(ref _entriesCount, -removedCount) > _flatModeThreshold)
+                {
+                    IsFlat = true;
+                }
+            }
+        }
+        finally
+        {
+            // Reset purging flag.
+            lock (_purgingLock)
+            {
+                _isPurging = false;
+            }
+        }
+    }
+
+    private int RemoveDeadKeys()
+    {
+        var removed = 0;
+        List<int> deadKeys = new();
+        ITaintedObject? previous;
+
+        foreach (var key in _map.Keys)
```

Comment:
I'm not sure if this is a good idea from a locking point of view? I can't remember how the locking works with concurrentDictionary, but either
- This is indeterminate, as keys can be added while enumerating
- It locks for the whole time.

Maybe it would be preferable to do a `ToArray()` on it first and iterate over that, as it's definitely safe?